tiny_dnn/core/framework/tensor.h

```diff
@@ -264,83 +347,49 @@ class Tensor {
+        return storage_pointer_->size();
     }
 
     void fill(U value) {
+        static_assert(!kConst, "Non-constant operation on constant Tensor");
         data_is_on_host_ = true;
         data_dirty_ = true;
-        std::fill(std::begin(host_data_), std::end(host_data_), value);
+        std::fill(std::begin(*storage_pointer_), std::end(*storage_pointer_), value);
     }
 
-
     //TODO(Randl): variadic template version of reshape
     void reshape(const std::array<size_t, kDimensions> &sz) {
+        static_assert(!kConst, "Non-constant operation on constant Tensor");
         shape_ = sz;
-        host_data_.resize(calcSize(), U(0));
+        storage_pointer_->resize(calcSize(), U(0));
     }
 
 private:
     size_t calcSize() const {
-        return std::accumulate(std::begin(shape_), std::end(shape_), size_t(1), std::multiplies<size_t>());
-    }
-
-    void toDevice() const {
-        if (data_is_on_host_ && data_dirty_) {
-#if defined(USE_OPENCL) || defined(USE_CUDA)
-            CLCudaAPI::Queue queue = device_->queue();
-            if (device_data_ && device_data_->GetSize() >= host_data_.size()) {
-                device_data_->Write(queue, host_data.size(), host_data_.data(), 0);
-            }
-            else {
-                CLCudaAPI::Context ctx = device_->context();
-                device_data_ = make_unique<CLCudaAPI::Buffer<U> >(
-                    ctx, queue, host_data_.begin(), host_data_.end());
-            }
-#endif
-            data_is_on_host_ = false;
-            data_dirty_ = false;
-        }
+        return std::accumulate(std::begin(shape_),
+                               std::end(shape_),
+                               size_t(1),
+                               std::multiplies<size_t>());
     }
 
-    void fromDevice() const {
-        if (!data_is_on_host_ && data_dirty_) {
-#if defined(USE_OPENCL) || defined(USE_CUDA)
-            assert(device_);
-            assert(device_data_);
-            device_data_->Read(device_->queue(), host_data_.size(), const_cast<U*>(host_data_.data())); // using const_cast<> to avoid making host_data_ entirely mutable
-#endif
-            data_is_on_host_ = true;
-            data_dirty_ = false;
-        }
-    }
-
-private:
     /**
      * A tensor holds data in C-style nD array, i.e row-major order:
      * the rightmost index “varies the fastest”.
      */
     std::array<size_t, kDimensions> shape_;
 
-    /* Pointer to the Tensor data in pure in the host device */
-    std::vector<U, aligned_allocator<U, 64>> host_data_;
-
-#if defined(USE_OPENCL) || defined(USE_CUDA)
-    /* Pointer to the Tensor data in the device */
-    std::unique_ptr<CLCudaAPI::Buffer<U> > device_data_;
-#endif
-    mutable bool data_is_on_host_;    //< current data is on host if true, on device if false.
-    mutable bool data_dirty_;         //< set to true if current data might have been modified
+    /* Offset from the beginningng of TensorStorage */
+    size_t offset;
```

Comment:
I think the rationale is to use "offset" for indexing subtensors.
E.G., if I have a 2x2x2 tensor, and I want to index the second in the first dimension, the subtensor will be a tensor class sharing the same storage, but with offset 4 (the size of the subtensor to be skipped to get to the element in the storage).